text/0000-mut-pattern-shorthand.md

```diff
@@ -0,0 +1,569 @@
+a feature like this when we have the time.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## Do nothing
+
+Of course, we could opt not to do this, in which case the ergonomic improvements
+would not be gained.
+
+## `rustfix`able `let mut (x, y)`
+
+One alternative to adding this in the language is to suggest `let (mut x, mut y)`
+whenever a user writes `let mut (x, y)`. But once we're on the path of doing
+this reliably, we might as well implement the whole feature or a subset of it
+and let the user write as they think.
+
+## Rationale for more than top level `mut`
+
+We could artificially take a more conservative approach,
+only permitting `let mut (x, (y, z)) = ..` and not `let (x, mut (y, z)) = ..`.
+However, this makes the feature less useful and does not make the grammar
+particularly simpler.
+
+## Rationale for more than `let` bindings
+
+Similarly, we could also only allow this for `let` bindings and not `if let`,
+`while let`, `for`, `match` and `fn` arguments. However, more uniformity does
+perhaps counter-intuitively help teachability and makes the hit to the
+complexity budget smaller.
+
+## Rationale for conservatism
+
+> When adding syntactic sugar, it’s usually best to try to fix things that
+> people try, but cannot do.
+>
+> Thinking ahead is good, but adding features just in case is not.
+
+*\- @varkor*
+
+We could opt to include the future work now.
+However, this RFC argues that we should not.
+
+The reasoning behind this is that while there are real world examples of tuples
+that would be made more ergonomic and readable, the RFC author could not find
+corresponding examples for structs and enum variants.
+
+Furthermore, the pattern `mut (x, y, ..)` is unambiguous more local
+than `mut Foo { bar, baz }`. While `mut (x, y, ..)` has been attempted by others
+and the author of this RFC before, the author is not aware of similar attempts
+for `Foo`.
+
+One could perhaps make the case that `mut Foo(x, y)` is different,
+but even so, we can always add that syntactic sugar later when there is a need
+for it. A problem with `mut Foo(x, y)` is that a newcomer might ask:
+*"How can `Foo` be mutable?"*. Indeed, this pattern would allow `mut Some(x)`
+instead of `Some(mut x)` which is less readable and local.
+
+### On the other hand - Macros
+
+One of the benefits listed in the [motivation] is that it the changes would
+make it easier for macro authors to introduce mutable bindings everywhere
+in a pattern. However, since this only applies to tuples, it also becomes
+less useful for macros. Therefore, one argument in favor of a more radical
+approach is to support macros better.
+
+# Prior art
+[prior-art]: #prior-art
+
+The combination of pattern matching and mutability is not particularly common.
+As such, the RFC's author is not aware of any prior art.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+None as of yet.
+
+# Future work
```

Comment:
We have recently implemented #554 (rust-lang/rust#48500). Would `mut` next to a parenthesis pattern be allowed?

```rust
    match 5 {
        mut (x) => {}
    }
```

Reply:
So looking as `parser-lalr.y`, [we have](https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L964): 

```ebnf
pat
: ...
| '(' pat_tup ')' { $$ = mk_node("PatTup", 1, $2); }
```

which I extended with `MUT '(' pat_tup ')'`
Then, [`pat_tup`](https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L1024) eventually leads to a production `MUT '(' pat ')'` so it seems that per the grammar I've included in the reference, `mut (x) => {}` would be permitted.

Is that a good thing or a bad thing..? I don't know. It seems simpler to permit `mut (x)` even if it is not the recommended style.

Reply:
@Centril the `*.y` file is not up-to-date though, and `(x)` isn't a tuple (the 1-tuple syntax is `($pat,)`).

Reply:
Ah damn :/ so my answer can't be reliable here if the `*.y` file is not up to date.

If we want to do the broader change per @scottmcm's comment above, it makes sense to also permit `mut (x)`. What is your view?

Reply:
@Centril I think if parenthesis is allowed, then almost all patterns should support leading mut (except the ambiguous-to-parse stuff like `x @ pat`, `x | y` and `x..=y`) because literally pattern can appear inside `(…)`.

Reply:
Hmm, @kennytm's comment pushes me back down to -0 on this, since it reminded me that the proposed grammar will allow all kinds of silliness like `mut (1...10, 20...30)`.  That's certainly not a blocking issue -- I'm confident it could be linted without too much trouble -- but does emphasize that the current "mut is part of a binding" rule is substantially simpler.

Reply:
So thinking about what such a lint could be, an analysis that checks the pattern `(1..=10, 20..=30)` for bindings would find that there are none, and so a `mut` in front of it would apply to zero bindings, which would be the lint condition. That's not so complex?

The current rule is for sure simpler tho =)